In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt


In [2]:


data = pd.read_csv(r"C:\Users\conno\workspace\projects\ufc_predictive_model\ufc-fighters-statistics.csv")

print(data.isnull().sum)

<bound method DataFrame.sum of        name  nickname   wins  losses  draws  height_cm  weight_in_kg  \
0     False      True  False   False  False      False         False   
1     False      True  False   False  False      False         False   
2     False     False  False   False  False      False         False   
3     False     False  False   False  False      False         False   
4     False      True  False   False  False      False         False   
...     ...       ...    ...     ...    ...        ...           ...   
4104  False     False  False   False  False       True         False   
4105  False      True  False   False  False      False         False   
4106  False      True  False   False  False       True         False   
4107  False      True  False   False  False      False         False   
4108  False     False  False   False  False      False         False   

      reach_in_cm  stance  date_of_birth  \
0            True    True           True   
1           Fals

In [3]:
data.head(10)

name                  nickname  wins  losses  draws  \
0           Rex Richards                       NaN     7       2      0   
1           Luana Santos                       NaN     6       1      0   
2     Cortavious Romious  Are You Not Entertained?     7       2      0   
3            Danny Silva                   El Puma     8       1      0   
4          Angel Pacheco                       NaN     7       2      0   
5  Charalampos Grigoriou                 Ferocious     8       3      0   
6            JR Coughran                   Bam Bam     4       1      0   
7     Ernesta Kareckaite              Heavy-Handed     5       0      1   
8           Danny Barlow              LeftHand2God     7       0      0   
9            Kim Couture                Sugar Free     3       8      0   

   height_cm  weight_in_kg  reach_in_cm    stance date_of_birth  \
0     195.58        120.20          NaN       NaN           NaN   
1     167.64         56.70       170.18  Orthodox    2000-04-16   
2     162.56         61.23       172.72  Southpaw    1994-01-06   
3     180.34         65.77       177.80    Switch    1997-01-30   
4     172.72         65.77       177.80  Orthodox    1992-01-13   
5     170.18         61.23       170.18  Orthodox    1992-03-31   
6     167.64         65.77          NaN  Orthodox    1994-05-17   
7     175.26         56.70       180.34  Orthodox    1998-07-05   
8     187.96         77.11       200.66  Southpaw    1995-08-02   
9     172.72         61.23          NaN  Orthodox    1975-12-18   

   significant_strikes_landed_per_minute  significant_striking_accuracy  \
0                                  17.65                           58.0   
1                                  17.10                           60.0   
2                                  14.48                           31.0   
3                                  13.60                           64.0   
4                                  13.13                           40.0   
5                                  13.00                           65.0   
6                                  12.29                           57.0   
7                                  12.27                           45.0   
8                                  12.15                           57.0   
9                                  12.12                           80.0   

   significant_strikes_absorbed_per_minute  significant_strike_defence  \
0                                     5.29                        72.0   
1                                     6.52                        68.0   
2                                    20.69                        41.0   
3                                    13.13                        59.0   
4                                    13.60                        35.0   
5                                     4.00                        63.0   
6                                     6.80                        45.0   
7                                    11.20                        42.0   
8                                     5.32                        61.0   
9                                     2.31                        69.0   

   average_takedowns_landed_per_15_minutes  takedown_accuracy  \
0                                      0.0                0.0   
1                                      0.0                0.0   
2                                      0.0                0.0   
3                                      2.0               66.0   
4                                      0.0                0.0   
5                                      0.0                0.0   
6                                      0.0                0.0   
7                                      1.0               50.0   
8                                      0.0                0.0   
9                                      0.0                0.0   

   takedown_defense  average_submissions_attempted_per_15_minutes  
0               0.0            

In [4]:
data.describe()

wins       losses        draws    height_cm  weight_in_kg  \
count  4109.000000  4109.000000  4109.000000  3812.000000   4022.000000   
mean     12.360672     5.722073     0.264298   178.234439     77.395405   
std       9.376023     5.104569     0.822483     8.889213     17.988389   
min       0.000000     0.000000     0.000000   152.400000     47.630000   
25%       7.000000     2.000000     0.000000   172.720000     65.770000   
50%      11.000000     4.000000     0.000000   177.800000     77.110000   
75%      17.000000     8.000000     0.000000   185.420000     83.910000   
max     253.000000    83.000000    11.000000   226.060000    349.270000   

       reach_in_cm  significant_strikes_landed_per_minute  \
count  2183.000000                            4109.000000   
mean    181.803729                               2.442100   
std      10.679990                               1.996586   
min     147.320000                               0.000000   
25%     175.260000                               0.830000   
50%     182.880000                               2.330000   
75%     190.500000                               3.600000   
max     213.360000                              17.650000   

       significant_striking_accuracy  significant_strikes_absorbed_per_minute  \
count                    4109.000000                              4109.000000   
mean                       35.553906                                 3.146878   
std                        20.405083                                 2.850083   
min                         0.000000                                 0.000000   
25%                        27.000000                                 1.550000   
50%                        40.000000                                 2.940000   
75%                        49.000000                                 4.240000   
max                       100.000000                                52.500000   

       significant_strike_defence  average_takedowns_landed_per_15_minutes  \
count                 4109.000000                              4109.000000   
mean                    42.666586                                 1.248537   
std                     22.323827                                 1.931608   
min                      0.000000                                 0.000000   
25%                     36.000000                                 0.000000   
50%                     50.000000                                 0.590000   
75%                     58.000000                                 1.930000   
max                    100.000000                                32.140000   

       takedown_accuracy  takedown_defense  \
count        4109.000000       4109.000000   
mean           26.296666         39.015332   
std            28.728565         34.457155   
min             0.000000          0.000000   
25%             0.000000          0.000000   
50%            22.000000         42.000000   
75%            45.000000         66.000000   
max           100.000000        100.000000   

       average_submissions_attempted_per_15_minutes  
count                                   4109.000000  
mean                                       2.442100  
std                                        1.996586  
min                                        0.000000  
25%                                        0.830000  
50%                                        2.330000  
75%                                        3.600000  
max                                       17.650000

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4109 entries, 0 to 4108
Data columns (total 18 columns):
 #   Column                                        Non-Null Count  Dtype  
---  ------                                        --------------  -----  
 0   name                                          4109 non-null   object 
 1   nickname                                      2255 non-null   object 
 2   wins                                          4109 non-null   int64  
 3   losses                                        4109 non-null   int64  
 4   draws                                         4109 non-null   int64  
 5   height_cm                                     3812 non-null   float64
 6   weight_in_kg                                  4022 non-null   float64
 7   reach_in_cm                                   2183 non-null   float64
 8   stance                                        3287 non-null   object 
 9   date_of_birth                                 2975 non-null   o

In [6]:
# Remove a warrning that doesn't apply
pd.options.mode.chained_assignment = None  # default='warn'

# Creating df to be transformed & dropping na's from the table
df = data.dropna(subset=['reach_in_cm', 'height_cm', 'stance', 'date_of_birth'])
# Convert 'date_of_birth' to an age
df['age'] = pd.to_datetime('now').year - pd.to_datetime(df['date_of_birth']).dt.year

# Dropping useless columns
df.drop(['name', 'nickname', 'date_of_birth'], axis=1, inplace=True)

median_wins = df['wins'].median()

df['total_fights'] = df['wins'] + df['losses'] + df['draws']

df['win_rate'] = df['wins'] / df['total_fights']

# Categorizng win rates
df['win_rate_cat'] = pd.cut(df['wins'],
                            bins=[-1, median_wins, df['wins'].max()], # Setting the first bin to -1 ensures that fighters w/ 0 wins are included
                            labels=['Low Win Rate', 'High Win Rate'])


In [7]:
print(df.isna().sum())

wins                                            0
losses                                          0
draws                                           0
height_cm                                       0
weight_in_kg                                    0
reach_in_cm                                     0
stance                                          0
significant_strikes_landed_per_minute           0
significant_striking_accuracy                   0
significant_strikes_absorbed_per_minute         0
significant_strike_defence                      0
average_takedowns_landed_per_15_minutes         0
takedown_accuracy                               0
takedown_defense                                0
average_submissions_attempted_per_15_minutes    0
age                                             0
total_fights                                    0
win_rate                                        0
win_rate_cat                                    0
dtype: int64


In [8]:
print(df['stance'].value_counts())

stance
Orthodox       1610
Southpaw        386
Switch          164
Open Stance       2
Name: count, dtype: int64


In [9]:
print(df.nunique())

wins                                             48
losses                                           25
draws                                             7
height_cm                                        22
weight_in_kg                                     40
reach_in_cm                                      27
stance                                            4
significant_strikes_landed_per_minute           612
significant_striking_accuracy                    70
significant_strikes_absorbed_per_minute         647
significant_strike_defence                       69
average_takedowns_landed_per_15_minutes         481
takedown_accuracy                                82
takedown_defense                                 90
average_submissions_attempted_per_15_minutes    612
age                                              38
total_fights                                     69
win_rate                                        245
win_rate_cat                                      2
dtype: int64

In [10]:
contingency_table = pd.crosstab(df['stance'], df['win_rate_cat'])
print(contingency_table)

win_rate_cat  Low Win Rate  High Win Rate
stance                                   
Open Stance              1              1
Orthodox               855            755
Southpaw               196            190
Switch                 111             53


In [11]:
from scipy.stats import chi2_contingency

In [12]:
chi2, p, dof, expected = chi2_contingency(contingency_table)

print(f"Chi-Square Statistic: {chi2}")
print(f"P-value: {p}")

Chi-Square Statistic: 14.459507714744722
P-value: 0.0023419488409906703


### Chi-Square Statistic

The Chi-Square Statistic of 14.4595 suggests that there is a deviation between the observed frequencies of fighters' win rates across different stances and what we would expect to see if there were no association between stance and win rate. The higher the Chi-Square value, the greater the deviation.

### P-value

The P-value is a measure of the probability that the observed data would have occurred by chance alone under the null hypothesis (which, in this case, states there is no association between the variables). A P-value of 0.00234 is less than the typical significance level of 0.05. This low P-value indicates that the probability of observing your data (or more extreme) under the null hypothesis is very low.

In [13]:
Q1 = df['wins'].quantile(0.25)
Q3 = df['wins'].quantile(0.75)
IQR = Q3 - Q1

lower_bound = max(0, Q1 - 1.5 * IQR)
upper_bound = Q3 + 1.5 * IQR

# Regressor Models

In [14]:
from sklearn.preprocessing import LabelEncoder

# Label encoder technique
label_encoder = LabelEncoder()
df['stance'] = label_encoder.fit_transform(df['stance'].astype(str))

In [15]:
d_df = pd.get_dummies(df)
d_df = df.drop(columns=['wins', 'losses', 'draws', 'win_rate_cat']) # Need to encode stance to numerical values

d_df.head()

height_cm  weight_in_kg  reach_in_cm  stance  \
1     167.64         56.70       170.18       1   
2     162.56         61.23       172.72       2   
3     180.34         65.77       177.80       3   
4     172.72         65.77       177.80       1   
5     170.18         61.23       170.18       1   

   significant_strikes_landed_per_minute  significant_striking_accuracy  \
1                                  17.10                           60.0   
2                                  14.48                           31.0   
3                                  13.60                           64.0   
4                                  13.13                           40.0   
5                                  13.00                           65.0   

   significant_strikes_absorbed_per_minute  significant_strike_defence  \
1                                     6.52                        68.0   
2                                    20.69                        41.0   
3                                    13.13                        59.0   
4                                    13.60                        35.0   
5                                     4.00                        63.0   

   average_takedowns_landed_per_15_minutes  takedown_accuracy  \
1                                      0.0                0.0   
2                                      0.0                0.0   
3                                      2.0               66.0   
4                                      0.0                0.0   
5                                      0.0                0.0   

   takedown_defense  average_submissions_attempted_per_15_minutes  age  \
1             100.0                                         17.10   24   
2               0.0                                         14.48   30   
3             100.0                                         13.60   27   
4              33.0                                         13.13   32   
5               0.0                                         13.00   32   

   total_fights  win_rate  
1             7  0.857143  
2             9  0.777778  
3             9  0.888889  
4             9  0.777778  
5            11  0.727273

In [16]:
# Creating model variables
X = d_df.drop('win_rate', axis=1)
y = d_df['win_rate']

In [17]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import ExtraTreesRegressor

X_train, X_test, y_train, y_test = train_test_split(X, y)

In [18]:
et_reg = ExtraTreesRegressor()
et_reg.fit(X_train, y_train)

ExtraTreesRegressor()

In [19]:
et_predictions = et_reg.predict(X_test)
et_predictions

array([0.73395554, 0.77832501, 0.82348915, 0.68333696, 0.69863504,
       0.68212165, 0.7097303 , 0.67158488, 0.65260458, 0.78356495,
       0.80046956, 0.7375024 , 0.68569381, 0.75748739, 0.65750457,
       0.69957967, 0.66878214, 0.74672633, 0.71797203, 0.6690835 ,
       0.64871352, 0.69648801, 0.73755454, 0.76123394, 0.69211307,
       0.73122216, 0.74012608, 0.78072998, 0.70229126, 0.71124386,
       0.71371024, 0.81234889, 0.67906377, 0.75083285, 0.72751458,
       0.75167367, 0.8381823 , 0.72350218, 0.65538567, 0.7706524 ,
       0.66348277, 0.68958947, 0.68463045, 0.72635973, 0.77726869,
       0.72020834, 0.77535264, 0.66583083, 0.71526572, 0.77442978,
       0.71549536, 0.76912248, 0.79383061, 0.79977408, 0.88945628,
       0.72283406, 0.73407671, 0.74918729, 0.84925692, 0.77897479,
       0.85045246, 0.78016891, 0.70944161, 0.81439617, 0.7325273 ,
       0.72226616, 0.7219385 , 0.66588501, 0.64183626, 0.69751325,
       0.91465745, 0.78212647, 0.69082467, 0.71213943, 0.69684

In [20]:
from sklearn.metrics import mean_squared_error, r2_score

print('r2 score: ', r2_score(y_test, et_predictions))
print('mean squared error: ', mean_squared_error(y_test, et_predictions))

r2 score:  0.30169223014476687
mean squared error:  0.008991104520348004


In [21]:
from sklearn.model_selection import cross_val_score

scores = cross_val_score(et_reg, X, y, cv=5)
print('cross validation scores: ', scores)

cross validation scores:  [0.19531617 0.25052703 0.2512854  0.16635535 0.16097489]


In [23]:
feature_importances = pd.DataFrame(zip(d_df.columns, et_reg.feature_importances_), columns=['feature', 'score'])

feature_importances = feature_importances.sort_values(by='score', ascending=False)

print(feature_importances)

                                         feature     score
12                                           age  0.196015
13                                  total_fights  0.081252
6        significant_strikes_absorbed_per_minute  0.080401
5                  significant_striking_accuracy  0.078848
8        average_takedowns_landed_per_15_minutes  0.075084
10                              takedown_defense  0.073445
7                     significant_strike_defence  0.062974
2                                    reach_in_cm  0.057238
11  average_submissions_attempted_per_15_minutes  0.054321
0                                      height_cm  0.054151
1                                   weight_in_kg  0.054118
4          significant_strikes_landed_per_minute  0.054052
9                              takedown_accuracy  0.046286
3                                         stance  0.031814


# Classifier Models

In [ ]:
# Creating model variables
X2 = df.drop('win_rate_cat', axis=1)
y2 = df['win_rate_cat']

X2_train, X2_test, y2_train, y2_test = train_test_split(X2, y2)

In [ ]:
from sklearn.preprocessing import LabelEncoder

# Label encoder technique
label_encoder = LabelEncoder()
df['stance'] = label_encoder.fit_transform(df['stance'].astype(str))

In [ ]:
# Descion Tree
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

dt_clf = DecisionTreeClassifier(random_state=42)
dt_clf.fit(X2_train, y2_train)
dt_predictions = dt_clf.predict(X2_test)

print(f"Decision Tree Accuracy: {accuracy_score(y2_test, dt_predictions)}")

In [ ]:
from sklearn.ensemble import RandomForestClassifier

rf_clf = RandomForestClassifier(random_state=42)
rf_clf.fit(X_train, y_train)
rf_predictions = rf_clf.predict(X_test)

print(f"Random Forest Accuracy: {accuracy_score(y_test, rf_predictions)}")

In [ ]:
from sklearn.ensemble import GradientBoostingClassifier

gb_clf = GradientBoostingClassifier(random_state=42)
gb_clf.fit(X_train, y_train)
gb_predictions = gb_clf.predict(X_test)

print(f"GBM Accuracy: {accuracy_score(y_test, gb_predictions)}")

### Results
This task is too simple as is.